### ISSUE TITLE
TV Video playback stalls while moving Kodi window in window mode, playback has to catch up afterwards

### ISSUE THREAD

**Issue by mhanor** [xbmc/xbmc#20851]
### Describe the bug
Moving the Kodi window (window mode) across the desktop, makes the TV video playback stall until the movement is stopped, at which point the playback has to catch up (speeds up the part that was delayed or playback hangs, audio included, with playback recovering after a while, if the initial delay was greater than, say, 10-20 seconds).
Also, during the movement of the window, the kodi.log file is flooded with continuously repeating lines, see below in the report.
This problem occurs with or without VAAPI enabled.

## Expected Behavior
No Video playback stalling.

## Actual Behavior
TV Video playback stalls while moving the Kodi window

## Possible Fix
Don't move the window across the desktop :smile: 

### To Reproduce
1. Start Kodi in window mode
2. Select TV
3. Select a TV channel to start playback
4. Move the Kodi window across the desktop and check the video content and the kodi.log file

### Debuglog
022-01-14 21:51:35.991 T:3045539 WARNING <general>: OutputPicture - timeout waiting for buffer
2022-01-14 21:51:35.995 T:3045433    INFO <general>: GL: Maximum texture width: 16384
2022-01-14 21:51:36.042 T:3045539    INFO <general>: Skipped 3 duplicate messages..
2022-01-14 21:51:36.042 T:3045539 WARNING <general>: OutputPicture - timeout waiting for buffer
2022-01-14 21:51:36.044 T:3045433    INFO <general>: GL: Maximum texture width: 16384
2022-01-14 21:51:36.094 T:3045539    INFO <general>: Skipped 4 duplicate messages..
The debuglog can be found [here](https://drive.google.com/file/d/1ga3Api-ukPjjASvq3S7pgwSK-tD8P5En/view?usp=sharing)

### Screenshots 
Not needed

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:
I think that window mode probably is not the usual mode of most Kodi users, but it would be nice to have a fix for this.

### Your Environment
Used Operating system:
 - [X] Linux

 - Operating system version/name:
Linux 5.15.0-2-amd64 1 SMP Debian 5.15.5-2 (2021-12-18) x86_64 GNU/Linux
Mate Desktop

 - Kodi version:
19.3 Debian package version: 2:19.3+dfsg1-1+b1 Media Center Kodi from Debian
Copyright (C) 2005-2021 Team Kodi from Debian - http://kodi.tv

**Comment by DaVukovic** [comment]
Not a debuglog

Partially unsupported Kodi version: `Starting Kodi from Debian (19.3 Debian package version`

stone-age GPU (10 years old): Mesa DRI Intel(R) HD Graphics 400 (BSW)

What do you expect?

@basilgello maybe something for you

Going to close this one as I'm not seeing an issue with that combination of hardware. If someone feels this is wrong, feel free to re-open it again.

**Comment by mhanor** [comment]
You say the GPU is old. But I have no issue while moving the VLC window during playback, or the Youtube out-of-window Firefox player. 
The same thing occurs on a I5 6600K with the Intel GPU (also an old GPU, but still...).

**Comment by enen92** [comment]
I gave this a go on my system (linux as well) and could not reproduce

**Comment by basilgello** [comment]
@DaVukovic It is a supported version :) Supported by me :)

@mhanor Sounds like a vaapi issue. One easy step is to check if VAAPI is at all enabled. Can you post the screenshot of video acceleration settings?

**Comment by mhanor** [comment]
@enen92 what Desktop environment are you using? Maybe it's relevant
@basilgello I've mentioned in the bug description that this occurs with or without VAAPI enabled in the Kodi settings
I will get back after I've installed Ubuntu on the other system and use the official Kodi.
I'm attaching a proper debug log (I hope I got it right this  time).
Thank you all for your feedback.
[kodi.log.zip](https://github.com/xbmc/xbmc/files/7874996/kodi.log.zip)
![Kodi VAAPI settings](https://user-images.githubusercontent.com/5107323/149622627-26b7a7ea-cf30-4d1b-8252-7ceaa8940ee9.png)
.

**Comment by enen92** [comment]
gnome 41, mesa-drivers

**Comment by DaVukovic** [comment]
@basilgello hence i said "partially" ;)

If you feel it's a real bug. Feel free to re-open it

**Comment by mhanor** [comment]
I was able to reproduce the issue with Ubuntu 21.10 with Mate Desktop+Xorg, but not with the default setup Gnome+Wayland (nor with Gnome+Xorg), on the Intel I5 6600K + HD530 GPU system, using Kodi official (stable) version:
19.3 (19.3.0) Git:20211024-49a04cd6a7 Media Center Kodi
Copyright (C) 2005-2021 Team Kodi - http://kodi.tv
[Debug log](https://github.com/xbmc/xbmc/files/7875633/kodi.log.2.zip)

I also tried it on a 3rd system (Ubuntu 21.10, GPU AMD Radeon HD6450), but it did not reproduce, neither on  Gnome+Wayland, nor on Mate+Xorg.

**Comment by basilgello** [comment]
> If you feel it's a real bug. Feel free to re-open it

Re-opening it after it becomes reproducible :)

**Comment by DaVukovic** [comment]
@basilgello Just to let you know. 

I'm not able to reproduce that. I've tested with Fedora 34, latest master, i5-6500 and a NVidia 1050ti. No matter if I turn hw accel on or off I don't have a stalled video if I move Kodi in windowed mode. 
The only thing I noticed is, while moving, Kodi takes about 90% of my CPU.

**Comment by mhanor** [comment]
I've installed Kodi nightly:
Kodi Media Center 20.0-ALPHA1 (19.90.101) Git:20220115-c69943bdcf
Copyright (C) 2005-2021 Team Kodi - http://kodi.tv

Moving the window and profiling the kodi process, with perf, on the Intel I5 6600K+HD530 system with Ubuntu 21.10, VAAPI is 
enabled, debug logging is disabled.

Gnome+Wayland -- no visible playback stalling, CPU usage appears normal
![perf profiling Gnome+Wayland](https://user-images.githubusercontent.com/5107323/149653311-25594b9f-e5a0-42ac-a862-3030e098f9ca.jpg)
Gnome+Xorg -- CPU usage is high, playback stalling is intermittent
![perf profiling Gnome+Xorg](https://user-images.githubusercontent.com/5107323/149653322-055e5201-9343-4c5f-8c02-a6aee98b71f6.jpg)
Mate+Xorg -- highest CPU usage, playback always stalls
![perf profiling Mate+Xorg](https://user-images.githubusercontent.com/5107323/149653330-21252ea3-3fdd-4b9f-aba2-072574b2e752.jpg)

**Comment by basilgello** [comment]
@mhanor Since you know how to profile things, can you run Kodi under gdb, wait for high CPU use and break with Ctrl-C then do `bt full` ? I remember such issue on my totally unrelated AMD laptop and what did it end with? The method above showed that the file thumbnail generator encountered a non-media file and stucked. I recalled that issue because why do I see a lot of XML reads on screen number 3?

**Comment by mhanor** [comment]
Here it is, gdb was attached to kodi post launch (if it matters):
https://paste.kodi.tv/dunugaguvu.kodi
I haven't found any info on how to get the kodi debug symbols files or server.

**Comment by basilgello** [comment]
@mhanor You should add Debian debug repository and install `kodi-bin-dbgsym` as described here: https://wiki.debian.org/HowToGetABacktrace

**Comment by basilgello** [comment]
Also I'd combine profiler and gdb to see whuch **thread** consumes the most CPU then look at thread map in gdb and see what it is

**Comment by mhanor** [comment]
Back to Debian testing then, running Debian Kodi version with debug symbols installed, on the Intel I5 6600K system.
Running top -H -p $pid reveals that all the action always occurs in thread 1, this is the thread that hogs the CPU. Also new views with perf report on recording made with --call-stack dwarf

While TV SD playback is ongoing:
https://paste.kodi.tv/upoxuhoruv.kodi

Playback stopped, OSD view only (TV channel list is in view)
https://paste.kodi.tv/juxajuqogo.kodi

I haven't saved the gdb bt output because I think the new perf views are much better than a one time snapshot of the thread's callstack, now that I understand better how to use it. I'm not a programmer, by the way, if I'm doing something wrong or you need something else, please let me know.

**Comment by basilgello** [comment]
@mhanor 

> I'm not a programmer, by the way, if I'm doing something wrong or you need something else, please let me know.

If more ppl were so ready to help like you, we'd forget about bugs hehe :) Thanks!

The shader getting recompiled on window resize is the culprit. I think we should redraw the scene only after resize end.
